File: sol/solution/0101-0200/0105/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder, inorder):
        self.idx = 0
        in_map = {val: i for i, val in enumerate(inorder)}
        def helper(left, right):
            if left > right:
                return None
            root_val = preorder[self.idx]
            self.idx += 1
            root = TreeNode(root_val)
            root.left = helper(left, in_map[root_val] - 1)
            root.right = helper(in_map[root_val] + 1, right)
            return root
        return helper(0, len(inorder) - 1)

File: sol/solution/0101-0200/0105/Solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
    idx := 0
    inMap := make(map[int]int)
    for i, v := range inorder {
        inMap[v] = i
    }
    var helper func(int, int) *TreeNode
    helper = func(left, right int) *TreeNode {
        if left > right {
            return nil
        }
        rootVal := preorder[idx]
        idx++
        root := &TreeNode{Val: rootVal}
        root.Left = helper(left, inMap[rootVal]-1)
        root.Right = helper(inMap[rootVal]+1, right)
        return root
    }
    return helper(0, len(inorder)-1)
}
